### PR TITLE
[Turbopack] HMR invalidation improvements

### PR DESCRIPTION
**Pull request by sokra**
### What?

* round chunk item sizes to cause less changes to chunking
* pick individual values from compile time info to avoid invalidating on change
* pick next config keys for more granular caching
* split completions into a tree

### CODE REVIEW

**Comment by sokra** [comment]
* **#69669** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69669?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#68896](https://github.com/vercel/next.js/pull/68896) <a href="https://app.graphite.dev/github/pr/vercel/next.js/68896?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#69060](https://github.com/vercel/next.js/pull/69060) <a href="https://app.graphite.dev/github/pr/vercel/next.js/69060?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#69668** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69668?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#70468** <a href="https://app.graphite.dev/github/pr/vercel/next.js/70468?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#69667** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69667?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#70494** <a href="https://app.graphite.dev/github/pr/vercel/next.js/70494?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @sokra and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: turbopack/crates/turbo-tasks/src/completion.rs [hunk 1]
```diff
@@ -60,14 +60,26 @@ impl Completions {
...
+                .iter()
+                .map(|&c| async move {
+                    c.await?;
+                    Ok(())
+                })
+                .try_join()
+                .await?;
+            Ok(Completion::new())
+        }
     }
```
Comment (sokra) [inline turbopack/crates/turbo-tasks/src/completion.rs:84]: Merging Completions in a tree

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 50761b5f5f5ed189d3d1141570b6f617118c04ef

`pnpm test-dev test/development/app-dir/server-components-hmr-cache/server-components-hmr-cache.test.ts`
- server-components-hmr-cache > edge runtime > in after() > should use cached fetch calls for fast refresh requests

<details>
<summary>Expand output</summary>

  ● server-components-hmr-cache › edge runtime › in after() › should use cached fetch calls for fast refresh requests

    expect(received).toEqual(expected) // deep equality

    Expected: "0.596027490003709"
    Received: "0.11017829313964733"

      73 |
      74 |         const valueAfterPatch = getLoggedAfterValue()
    > 75 |         expect(valueBeforePatch).toEqual(valueAfterPatch)
         |                                  ^
      76 |       })
      77 |
      78 |       it('should not use cached fetch calls for intentional refresh requests', async () => {

      at Object.toEqual (development/app-dir/server-components-hmr-cache/server-components-hmr-cache.test.ts:75:34)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: crates/next-core/src/next_server/transforms.rs [hunk 2]
```diff
@@ -32,11 +32,11 @@ pub async fn get_next_server_transforms_rules(
...
     let mut rules = vec![];
 
-    let modularize_imports_config = &next_config.await?.modularize_imports;
+    let modularize_imports_config = &next_config.modularize_imports().await?;
     let mdx_rs = next_config.mdx_rs().await?.is_some();
-    if let Some(modularize_imports_config) = modularize_imports_config {
+    if !modularize_imports_config.is_empty() {
         rules.push(get_next_modularize_imports_rule(
-            modularize_imports_config,
+            &modularize_imports_config,
```
Comment (mischnic) [inline crates/next-core/src/next_server/transforms.rs:39]:
```suggestion
            modularize_imports_config,
```

File: turbopack/crates/turbopack-core/src/chunk/mod.rs [hunk 3]
```diff
@@ -713,6 +713,18 @@ pub trait ChunkType: ValueToString {
...
 
+pub fn round_chunk_item_size(size: usize) -> usize {
+    if size == 0 {
+        return 0;
+    }
+    let a = size.next_power_of_two();
+    if a == size {
+        size
+    } else {
+        size & (a | (a >> 1) | (a >> 2))
```
Comment (mischnic) [inline turbopack/crates/turbopack-core/src/chunk/mod.rs:724]: So the idea here is to have exponentially growing spacing between sizes (just like `next_power_of_two`), but just slightly smaller spacing?

File: crates/next-core/src/next_client/transforms.rs [hunk 4]
```diff
@@ -29,11 +29,11 @@ pub async fn get_next_client_transforms_rules(
...
     let mut rules = vec![];
 
-    let modularize_imports_config = &next_config.await?.modularize_imports;
+    let modularize_imports_config = &next_config.modularize_imports().await?;
     let enable_mdx_rs = next_config.mdx_rs().await?.is_some();
-    if let Some(modularize_imports_config) = modularize_imports_config {
+    if !modularize_imports_config.is_empty() {
         rules.push(get_next_modularize_imports_rule(
-            modularize_imports_config,
+            &modularize_imports_config,
```
Comment (mischnic) [inline crates/next-core/src/next_client/transforms.rs:36]:
```suggestion
            modularize_imports_config,
```

File: turbopack/crates/turbopack-core/src/chunk/mod.rs (hunk 3 above)
Comment (sokra) [reply turbopack/crates/turbopack-core/src/chunk/mod.rs:724]: Yep. `next_power_of_two` would have a new size every time the value doubles.
This implementation has a new size every time the value increases by 50%

**Review by mischnic — APPROVED** [review]

**Review by bgw — APPROVED** [review]

**Comment by sokra** [comment]
### Merge activity

* **Oct 2, 5:27 AM EDT**: @sokra merged this pull request with [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/70494).

**Pull request merged**